### PR TITLE
Implement Spotify recommendations API for fallback songs with lyrics

### LINKED ISSUE #68: [BE] Integrate Spotify Recommendations API for no-lyrics fallback
Extend SpotifyService to call the Spotify Recommendations API seeded with a given spotifyId, then filter results to only those where lyrics.ovh returns lyrics.

Endpoint: GET https://api.spotify.com/v1/recommendations?seed_tracks={spotifyId}&limit=10

Filter: run parallel lyrics.ovh checks on all results and keep only those with lyricsAvailable = true.

**Acceptance criteria:**
- [ ] Recommendations are fetched from Spotify using the seed track
- [ ] Results are filtered via parallel lyrics.ovh checks
- [ ] Returns up to 5 alternatives with lyricsAvailable = true

### LINKED ISSUE #69: [BE] Implement GET /songs/{id}/recommendations endpoint
Expose a REST endpoint that returns alternative song suggestions for a track that has no lyrics.

Endpoint: GET /songs/{spotifyId}/recommendations
Response: list of up to 5 songs, same shape as /songs/search results, all with lyricsAvailable = true

**Acceptance criteria:**
- [ ] Returns recommendations seeded by the given spotifyId
- [ ] All returned songs have lyricsAvailable = true
- [ ] Returns 404 if spotifyId is invalid or no recommendations found

### PR DESCRIPTION
**Pull request by doemefu**
This pull request introduces a new feature for song recommendations based on a given Spotify track, ensuring that only tracks with available lyrics are returned. It also adds robust fallback logic to the song search flow, improves service and controller structure, and includes comprehensive unit tests for the new and updated behaviors.

**Feature: Song Recommendations with Lyrics**

* Adds a new endpoint `GET /songs/{spotifyId}/recommendations` to return up to 5 recommended tracks with `lyricsAvailable=true`, or a 404 if none are found. [[1]](diffhunk://#diff-be991424fed6a11f0c5ca87c4fcecbaa76795336e5399d769c47037dbab50e6fR9-R10) [[2]](diffhunk://#diff-be991424fed6a11f0c5ca87c4fcecbaa76795336e5399d769c47037dbab50e6fR92-R99)
* Implements `getRecommendationsForSong` in `SongService`, which fetches recommendations from Spotify and filters for tracks with lyrics, throwing a 404 if none are available.

**Enhancements to Song Search Logic**

* Updates the `search` method in `SongService` to trigger a recommendations fallback: if all search results lack lyrics, it returns the first result plus up to 3 recommendations with lyrics, iterating as needed to meet a minimum.

**Spotify Integration Improvements**

* Adds `getRecommendations` to `SpotifyService`, integrating with the Spotify Recommendations API and handling authentication retries. [[1]](diffhunk://#diff-cda8e0d17d5df2b8f2895d6188bb5d8e2f1753b64dcb9ceebfcf93b8450ed4eeR26-R27) [[2]](diffhunk://#diff-cda8e0d17d5df2b8f2895d6188bb5d8e2f1753b64dcb9ceebfcf93b8450ed4eeR90-R133)

**Testing and Validation**

* Adds/updates controller and service tests to cover the recommendations endpoint, fallback logic, and edge cases (e.g., no lyrics found, result limiting, filtering). [[1]](diffhunk://#diff-c5bcbd4b361684985fddbf0d8fb5ec2ddfefdb499fbe824c55cc323b30a44c83R117-R145) [[2]](diffhunk://#diff-ee464a2a6999603cf8dac2a66de11f2a43a4d021beb95c2a9f81c725783fa249R252-R370) [[3]](diffhunk://#diff-0aeaaa00e77e726f0b66909fab2c97dd0198ccc2f5e5e6c3a48bff96126246efR141-R183) [[4]](diffhunk://#diff-ee464a2a6999603cf8dac2a66de11f2a43a4d021beb95c2a9f81c725783fa249R73-R74) [[5]](diffhunk://#diff-ee464a2a6999603cf8dac2a66de11f2a43a4d021beb95c2a9f81c725783fa249R98-R99)

**Dependency and Import Adjustments**

* Updates imports and dependencies to support new features and tests. [[1]](diffhunk://#diff-843b2a0ff682f99ed650f11c6640b7f4a5784bd1cd3a768ada305d671ea368bcR11-R22) [[2]](diffhunk://#diff-c5bcbd4b361684985fddbf0d8fb5ec2ddfefdb499fbe824c55cc323b30a44c83R21)

These changes collectively improve the user experience by ensuring that recommended songs always have lyrics available and that the search experience degrades gracefully when lyrics are missing.

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java [hunk 1]
```diff
@@ -44,37 +49,100 @@ public SongService(SpotifyService spotifyService, LyricsService lyricsService,
...
+        int lyricsCount = 0;
+        int attempts = 0;
+        while (lyricsCount < minWithLyrics && attempts < 3) {
+            List<SpotifyTrack> batch = spotifyService.getRecommendations(seedTrackId);
+            for (SongSearchResultDTO dto : checkLyricsAndBuildDtos(batch)) {
+                if (seenIds.add(dto.getSpotifyId())) {
+                    collected.add(dto);
+                    if (Boolean.TRUE.equals(dto.getLyricsAvailable())) lyricsCount++;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java:138]: Fixed. The batch is now pre-filtered by `seenIds` before being passed to `checkLyricsAndBuildDtos`, so lyrics are only fetched for tracks not already seen in previous rounds.

File: src/main/java/ch/uzh/ifi/hase/soprafs26/controller/SongsController.java [hunk 2]
```diff
@@ -87,4 +89,12 @@ public ResponseEntity<Void> sessionsSessionIdSongsSongIdDelete(Long sessionId, L
     public ResponseEntity<Void> sessionsSessionIdSongsSongIdPlayedPut(Long sessionId, Long songId) {
         return ResponseEntity.noContent().build();
     }
+
+    // GET /songs/{spotifyId}/recommendations — Issue #69
+    // Returns up to 5 recommendations with lyricsAvailable=true, 404 if none found
+    @GetMapping("/songs/{spotifyId}/recommendations")
+    public ResponseEntity<List<SongSearchResultDTO>> songRecommendationsGet(
+            @PathVariable String spotifyId) {
+        return ResponseEntity.ok(songService.getRecommendationsForSong(spotifyId));
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/uzh/ifi/hase/soprafs26/controller/SongsController.java:98]: Fixed. Added `/songs/{spotifyId}/recommendations` to `karaokee-openapi.json` with path parameter, 200 (array of SongSearchResultDTO) and 404 responses.

File: src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java [hunk 3]
```diff
@@ -44,37 +49,100 @@ public SongService(SpotifyService spotifyService, LyricsService lyricsService,
...
+            attempts++;
+        }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java:140]: The iteration is intentional per the feature requirement ('iterate until we have at least two songs with lyrics'). Spotify's recommendations API is probabilistic and does return different tracks across calls with the same seed. The `seenIds` HashSet (now applied before the lyrics lookup) ensures duplicates from previous rounds are skipped efficiently.

File: src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java [hunk 4]
```diff
@@ -44,37 +49,100 @@ public SongService(SpotifyService spotifyService, LyricsService lyricsService,
...
+    /**
+     * Returns up to 5 Spotify recommendations for the given seed track that have lyrics available.
+     * Throws 404 if no recommendations with lyrics are found.
+     */
+    public List<SongSearchResultDTO> getRecommendationsForSong(String spotifyId) {
+        List<SpotifyTrack> tracks = spotifyService.getRecommendations(spotifyId);
+        List<SongSearchResultDTO> withLyrics = checkLyricsAndBuildDtos(tracks).stream()
+                .filter(SongSearchResultDTO::getLyricsAvailable)
+                .limit(5)
+                .toList();
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/uzh/ifi/hase/soprafs26/service/SongService.java:112]: Not changing this. Lyrics are fetched in parallel via `CompletableFuture.allOf`, so fetching 10 tracks costs the same wall-clock time as fetching 5. Sequential early-exit would be slower. The extra concurrent requests to lyrics.ovh are the deliberate tradeoff for faster response times.

File: src/main/java/ch/uzh/ifi/hase/soprafs26/service/SpotifyService.java [hunk 5]
```diff
@@ -85,6 +87,50 @@ public List<SpotifyTrack> search(String query) {
         }
     }
 
+    public List<SpotifyTrack> getRecommendations(String spotifyId) {
+        try {
+            return doGetRecommendations(spotifyId);
+        }
+        catch (HttpClientErrorException.Unauthorized e) {
+            log.warn("Spotify token rejected (401) — refreshing and retrying once");
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/uzh/ifi/hase/soprafs26/service/SpotifyService.java:95]: This matches the existing pattern in `search()`, which also only handles 401. Non-401 errors from Spotify propagate as-is, which is consistent across the service. The `GlobalExceptionAdvice` handles `HttpServerErrorException` at the application boundary.
